pkg/proxy/proxy.go

```diff
@@ -136,130 +145,191 @@ func (p *Proxy) Run(stopCh <-chan struct{}) (<-chan struct{}, error) {
 	return waitCh, nil
 }
 
-func (p *Proxy) serve(proxyHandler *httputil.ReverseProxy, stopCh <-chan struct{}) (<-chan struct{}, error) {
+func (p *Proxy) serve(handler http.Handler, stopCh <-chan struct{}) (<-chan struct{}, error) {
+	// Setup proxy handlers
+	handler = p.withHandlers(handler)
+
 	// securely serve using serving config
-	waitCh, err := p.secureServingInfo.Serve(proxyHandler, time.Second*60, stopCh)
+	waitCh, err := p.secureServingInfo.Serve(handler, time.Second*60, stopCh)
 	if err != nil {
 		return nil, err
 	}
 
 	return waitCh, nil
 }
 
+func (p *Proxy) withHandlers(handler http.Handler) http.Handler {
+	// Set up proxy handlers
+	handler = p.withImpersonateRequest(handler)
+	handler = p.withAuthenticateRequest(handler)
+	return handler
+}
+
+// RoundTrip is called last and is used to manipulate the forwarded request using context.
 func (p *Proxy) RoundTrip(req *http.Request) (*http.Response, error) {
-	// Clone the request here since successfully authenticating the request
-	// deletes those auth headers
-	reqCpy := utilnet.CloneRequest(req)
+	// Here we have successfully authenticated so now need to determine whether
+	// we need use impersonation or not.
+
+	// If no impersonation then we return here without setting impersonation
+	// header but re-introduce the token we removed.
+	if context.NoImpersonation(req.Context()) {
+		token := context.BearerToken(req.Context())
+		req.Header.Add("Authorization", token)
+		return p.noAuthClientTransport.RoundTrip(req)
+	}
 
-	// auth request and handle unauthed
-	info, ok, err := p.oidcRequestAuther.AuthenticateRequest(reqCpy)
-	if err != nil {
+	// Get the impersonation headers from the context.
+	conf := context.ImpersonationConfig(req.Context())
+	if conf == nil {
+		return nil, errNoImpersonationConfig
+	}
 
-		// attempt to passthrough request if valid token
-		if p.options.TokenReview {
-			return p.tokenReview(reqCpy)
-		}
+	// Set up impersonation request.
+	rt := transport.NewImpersonatingRoundTripper(*conf, p.clientTransport)
 
-		return nil, errUnauthorized
-	}
+	// Push request through round trippers to the API server.
+	return rt.RoundTrip(req)
+}
 
-	// failed authorization
-	if !ok {
-		return nil, errUnauthorized
-	}
+// withAuthenticateRequest adds the proxy authentication handler to a chain.
+func (p *Proxy) withAuthenticateRequest(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		// Auth request and handle unauthed
+		info, ok, err := p.oidcRequestAuther.AuthenticateRequest(req)
+		if err != nil {
+			if !p.config.TokenReview {
+				p.handleError(rw, req, errUnauthorized)
+				return
+			}
+
+			// Attempt to passthrough request if valid token
+			if p.reviewToken(rw, req) {
```

Comment:
Should we check what the actual error is before attempting an alternate authentication strategy here? What sort of errors are genuine errors and require an immediate return, and what sort are 'acceptable' and indicate the request should be delegated elsewhere?

Reply:
Both a network or auth fail error are treated the same here.

If the token fails review then it obviously doesn't auth.

If the network fails then we are unable to validate the token and therefore treat as the same unauth.

Not sure if we should be doing any retries here personally.

Reply:
Context: 
```
if err == nil && ok {
		klog.V(4).Infof("passing request with valid token through (%s)",
			req.RemoteAddr)

		return true
	}

	if err != nil {
		klog.Errorf("unable to authenticate the request via TokenReview due to an error (%s): %s",
			req.RemoteAddr, err)
	}
```